### ISSUE TITLE
Truncating headers file at height throws ValueError: Invalid mode ('rwb+')

### ISSUE THREAD

**Issue by shsmith** [spesmilo/electrum#2335]
I set a checkpoint on the main testnet chain and after that launching electrum fails with this traceback:
```
python electrum -v --testnet --segwit
[SimpleConfig] electrum directory AppData\Roaming\Electrum\testnet
[Plugins] registering hardware digitalbitbox: ('hardware', 'digitalbitbox', u'Digital Bitbox wallet')
[Plugins] registering hardware keepkey: ('hardware', 'keepkey', u'KeepKey wallet')
[Plugins] registering hardware ledger: ('hardware', 'ledger', u'Ledger wallet')
[Plugins] registering hardware trezor: ('hardware', 'trezor', u'TREZOR wallet')
[Plugins] registering wallet type ('2fa', 'trustedcoin')
[profiler] __init__ 0.0100
[Blockchain] Truncating headers file at height 1096499
Traceback (most recent call last):
  File "electrum", line 377, in <module>
    d = daemon.Daemon(config, fd)
  File "electrum\lib\daemon.py", line 111, in __init__
    self.network = Network(config)
  File "electrum\lib\network.py", line 207, in __init__
    self.blockchain = Blockchain(self.config, self)
  File "electrum\lib\blockchain.py", line 42, in __init__
    self.check_truncate_headers()
  File "electrum\lib\blockchain.py", line 194, in check_truncate_headers
    f = open(name, 'rwb+')
ValueError: Invalid mode ('rwb+')
[Plugins] stopped
```

**Comment by shsmith** [comment]
Here is the checkpoint from my config file:
    "checkpoint_height": 1096499, 
    "checkpoint_value": "00000000f582579c7a7dadaf7558df9130639a76907449674e3cec83d899dba0",

**Comment by shsmith** [comment]
I deleted blockchain_headers and after a few relaunches it started looping on this:
```
[Network] connecting to testnet.hsmiths.com:53011:t as new interface
[testnet.hsmiths.com] connected
[Network] switching to testnet.hsmiths.com:53011:t
[Network] sending subscriptions to testnet.hsmiths.com:53011:t 0 80
[Network] fee_estimates[25] 20077
[Network] fee_estimates[10] 20077
[Network] fee_estimates[5] 20077
[Network] fee_estimates[2] 20077
[Network] relayfee 1000
[profiler] on_update 0.6750
[Blockchain] validated checkpoint 1096499
[Blockchain] validated chunk 543
[testnet.hsmiths.com] requesting chunk 543
[Blockchain] verify_chunk failed failed checkpoint
[testnet.hsmiths.com] requesting chunk 542
[profiler] on_update 0.8960
[Blockchain] validated chunk 542
[testnet.hsmiths.com] requesting chunk 543
[testnet.hsmiths.com] {u'message': u'chunk indices not advancing (wrong network?): [543, 542, 543]', u'code': -32600}
[testnet.hsmiths.com] connection closed remotely

```
Maybe in this situation the client should switch servers, or at least stop auto-connecting to a server that has failed the checkpoint.

This seems related to https://github.com/spesmilo/electrum/issues/2314

**Comment by ecdsa** [comment]
are you on Windows? does it work with 'rw+' ?

**Comment by shsmith** [comment]
>I changed the mode to rb+ ; that should fix the first issue

Confirmed. Truncate is successfully now.

Also seemed to switch servers on failed checkpoint:
```
[Blockchain] verify_chunk failed failed checkpoint
[testnet.hsmiths.com] requesting chunk 542
[Blockchain] validated chunk 542
[testnet.hsmiths.com] requesting chunk 543
[testnet.hsmiths.com] connection closed remotely
[testnet.hsmiths.com] {u'message': u'chunk indices not advancing (wrong network?): [543, 542, 543]', u'code': -32600}
[14.3.140.101] requesting header 1096523
[Network] switching to 14.3.140.101:51001:t
[Network] sending subscriptions to 14.3.140.101:51001:t 1 80
[14.3.140.101] requesting header 1096522
[profiler] on_update 2.3310
[Synchronizer] receiving history mxbjbzKtCDcZWGQryAJpPTq3HTUZuYxtWb 0
[Synchronizer] error: status mismatch: mxbjbzKtCDcZWGQryAJpPTq3HTUZuYxtWb
```
It then requested a number of headers until finally 'validated checkpoint'.
As soon as that finished, it switched back to the wrong side of fork and started requesting chunks again.
```
[Synchronizer] receiving history mxbjbzKtCDcZWGQryAJpPTq3HTUZuYxtWb 0
[Synchronizer] error: status mismatch: mxbjbzKtCDcZWGQryAJpPTq3HTUZuYxtWb
[Network] fee_estimates[25] 1012
[Network] fee_estimates[10] 1012
[Network] fee_estimates[5] 1012
[Network] fee_estimates[2] 1012
[Network] relayfee 0
[profiler] write 0.0000
[profiler] save_transactions 0.0510
[ElectrumWindow/default_wallet] Notifying GUI
[14.3.140.101] requesting header 1096521
[14.3.140.101] requesting header 1096520
  ...
[14.3.140.101] requesting header 1096500
[14.3.140.101] requesting header 1096499
[Blockchain] reorg
[14.3.140.101] requesting header 1096498
[Blockchain] reorg
[14.3.140.101] requesting header 1096497
[Blockchain] reorg
  ...
[Blockchain] reorg
[14.3.140.101] requesting header 1096419
[Blockchain] reorg
[14.3.140.101] requesting header 1096418

[Blockchain] validated checkpoint 1096499
[Blockchain] new height: 1096523

[Network] network: retrying connections
[testnet.hsmiths.com] connected
[testnet.hsmiths.com] requesting chunk 543
[Blockchain] verify_chunk failed failed checkpoint
[testnet.hsmiths.com] requesting chunk 542
[Blockchain] validated chunk 542
[testnet.hsmiths.com] requesting chunk 543
[testnet.hsmiths.com] {u'message': u'chunk indices not advancing (wrong network?): [543, 542, 543]', u'code': -32600}
[testnet.hsmiths.com] connection closed remotely
```

**Comment by shsmith** [comment]
Not sure if this is related or something new, but after "header didn't connect" I get a "bad file descriptor" and the process seems hung.
```
[testnet.hsmiths.com] requesting header 1100685
[Blockchain] failed checkpoint
[testnet.hsmiths.com] header didn't connect, dismissing interface

[14.3.140.101] requesting chunk 545
Exception in thread Thread-3:
Traceback (most recent call last):
  File "lib\threading.py", line 801, in __bootstrap_inner
    self.run()
  File "electrum\lib\network.py", line 859, in run
    self.wait_on_sockets()
  File " electrum\lib\network.py", line 844, in wait_on_sockets
    rout, wout, xout = select.select(rin, win, [], 0.1)
  File " electrum\lib\interface.py", line 243, in fileno
    return self.socket.fileno()
  File " lib\socket.py", line 228, in meth
    return getattr(self._sock,name)(*args)
  File " lib\socket.py", line 174, in _dummy
    raise error(EBADF, 'Bad file descriptor')
error: [Errno 9] Bad file descriptor
```

**Comment by ecdsa** [comment]
fixed in d371e1c807f3cf8b0a65d03f470f104baf5416cc
